Advanced diplotype clustering

sanjaynagi opened the pull request:
Builds on #507 and partially addresses #361. 

In this PR, I add an advanced diplotype clustering function which performs diplotype clustering on a given region, and plots alongside it sample heterozygosity and amino acid alleles. This is particularly useful for investigating resistance mutations at IR loci. 

It adds a few new methods:

- an `aa_individual_allele_frequencies()` function, which just runs aa_allele_frequencies but calculates within individual frequencies (0, 0.5 or 1) rather than in cohorts. It could be that transforming the SNP data directly would be more appropriate and potentially faster than this. It might make more sense to use counts of alleles (0,1,2). 
- a function to calculate and plot heterozygosity over a provided region (current methods in malariagen_data, `_sample_count_het()` are windowed and applied to one sample at a time)
- A function to concatenate plotly subplots into one figure (`concat_subplots()`)
- A function to extract the sample order from the dendrogram 

For now, I have not added CNV support, however, it should be added prior to merging this PR.

sanjaynagi commented:
ahhh, it seems I forgot I was working on a fork, as such, the google credentials dont work.  Whats the best approach in future? @ahernank @alimanfoo

ahernank commented:
Hi @sanjaynagi, a short-term solution, but you have permissions to write directly on the repo rather than using branches.

sanjaynagi commented:
I've added CNV support, though currently the bar is copy number, but potentially it should be # extra copies.

alimanfoo commented:
Thanks @sanjaynagi. Just to mention I updated this branch to bring in some unrelated test fixes from master.

alimanfoo commented:
Very exciting, I'll give this a proper look asap!

alimanfoo commented:
Hi @sanjaynagi, some investigation into the CNV plotting issue here: https://github.com/sanjaynagi/malariagen-data-python/pull/1

sanjaynagi commented:
Some updates:

thanks to @alimanfoo for sorting the CNV colorscale problem. 

Ive completely removed any usage of px.imshow() now and replaced with go.Heatmap() for heterozygosity and amino acid heatmaps as well as CNVs. This also simplifies the code further, as it seems that we now don't need to pass the x_range through various functions which we did previously. 

I've added a check to see if there are actually any amino acids above the MAF threshold, if there are not, an amino acid heatmap will not be created.

Think we just need to maybe write some tests, and a final check from @alimanfoo, then we may be good to go! yeehaw!! Ill have a go at zooming in the x-axis aswell (currently its slightly zoomed out). 

![image](https://github.com/malariagen/malariagen-data-python/assets/34922269/4067a360-181b-40ec-8f96-014d4fbe623d)

alimanfoo commented:
Hi @sanjaynagi, some comments and suggestions over in https://github.com/sanjaynagi/malariagen-data-python/pull/2

sanjaynagi commented on malariagen_data/anoph/dipclust.py:
```diff
@@ -273,3 +280,418 @@ def _diplotype_pairwise_distances(
+            plot_bgcolor="white",
+        )
+
+        # COMMENT: this feels overly complex - can we somehow remove the colorbar earlier,
```
Not sure - I did try that originally, but was struggling a lot to do so. Also, I think it requires making Figure objects for the heatmaps first, whereas currently we just make heatmap traces, so it kind of adds more code.

sanjaynagi replied:
haven't got time right now to implement the other stuff but feel free to do so

sanjaynagi commented on malariagen_data/anoph/dipclust.py:
```diff
@@ -273,3 +280,418 @@ def _diplotype_pairwise_distances(
         )
+
+    def _dipclust_extract_dendro_sample_order(self, fig):
+        # COMMENT: Is there a simpler way to obtain the sample order?
```
something tells me there is a simpler way  😂

sanjaynagi commented on malariagen_data/anoph/dipclust.py:
```diff
@@ -273,3 +280,418 @@ def _diplotype_pairwise_distances(
+        fig_dendro = res["figure"]
+        n_snps = res["n_snps"]
+        dendro_sample_id_order = (
+            # COMMENT: Can we simplify? Why do this sorting here?
```
something tells me we can 😂

sanjaynagi commented:
@alimanfoo I've had a go at all the comments except the removing colorbars earlier one, cos I tried to do that in the past and it took me a while and I couldn't :)

alimanfoo commented:
Have pushed a few suggested changes directly to this branch, hope that's OK. Still reviewing so may do a bit more.

alimanfoo commented:
Hi @sanjaynagi, I pushed a few more changes, includes:

* Set colorscale visibility when heatmaps are created.
* Calculate denominator for heterozygosity based on number of called genotypes.
* Remove logic for drawing lines on the SNP heatmap - that wasn't being used after recent refactoring.
* Renamed a few things from "aa_..." to "snp_..." to reflect that we are generalising to plotting any kind of SNP.

alimanfoo commented:
Hi @sanjaynagi, I think this is basically good to merge. Because this is a PR from a fork the CI tests won't run, so I'm thinking to merge this PR then open up a follow on PR from a branch within the main repo to confirm tests all run. How does that sound? Anything else you'd like to do here before I merge?

sanjaynagi commented:
Hey @alimanfoo . Sounds good to me. Nothing else to add currently.
